### PR TITLE
Feature: add delete project

### PR DESCRIPTION
**Pull request by Rahzbob**

### CODE REVIEW

File: tests/unit/pages/HomePage.spec.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+    ProjectEvent.DeleteProjectEvent.subscribe.mockImplementation(() => {
+      subscribe();
+      return { unsubscribe };
+    });
+
+    wrapper = shallowMount(HomePage);
+  });
+
+  describe('Test function: setProjects', () => {
```
Comment (Zorin95670) [inline tests/unit/pages/HomePage.spec.js:38]: same as before

**Review by Zorin95670 — CHANGES_REQUESTED** [review]

File: src/pages/HomePage.vue [hunk 2]
```diff
@@ -14,18 +14,46 @@
...
 import ImportProjectDialog from 'components/dialog/ImportProjectDialog';
+import DeleteProjectDialog from 'components/dialog/DeleteProjectDialog';
+import ProjectEvent from 'src/composables/events/ProjectEvent';
+import {
+  ref,
+  onMounted,
+  onUnmounted,
+} from 'vue';
+
+const projects = ref(getProjects());
```
Comment (justinehell) [inline src/pages/HomePage.vue:38]: suggestion 
`const projects = ref({});`

File: cypress/e2e/HomePage/delete.project.feature [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    And I visit the "/"
+    And I click on "[data-cy=\"new-project\"]"
+    And I expect current url is "modelizer/project-[a-f0-9]{8}/model"
+    Then I click on "[data-cy=\"app-logo-link\"]"
+    And I expect current url is "/"
+
+    When I click on "[data-cy=\"delete-project\"]"
+    And I click on "[data-cy=\"delete-project-form\"] [data-cy=\"confirm-delete-project-checkbox\"]"
+    And I click on "[data-cy=\"delete-project-form\"] [data-cy=\"delete-project-form-submit\"]"
+
```
Comment (justinehell) [inline cypress/e2e/HomePage/delete.project.feature:14]: remove empty line

File: cypress/e2e/HomePage/delete.project.feature [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    Then I expect "positive" toast to appear with text "Project has been deleted 😥"
+    And I expect "[data-cy=\"delete-project-form\"]" is closed
+    And I expect localstorage field "projects" is '{}'
```
Comment (justinehell) [inline cypress/e2e/HomePage/delete.project.feature:17]: do not check localstorage, but check that the deleted project card no longer exists

File: cypress/e2e/HomePage/delete.project.feature [hunk 5]
```diff
@@ -0,0 +1,17 @@
+Feature: Test homepage: project deletion
+
+  Scenario: Delete existing project
+    Given I clear cache
+    And I visit the "/"
+    And I click on "[data-cy=\"new-project\"]"
+    And I expect current url is "modelizer/project-[a-f0-9]{8}/model"
```
Comment (justinehell) [inline cypress/e2e/HomePage/delete.project.feature:7]: jump line
`When I click on...`
`Then I expect current url...`

File: cypress/e2e/HomePage/delete.project.feature [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    Then I click on "[data-cy=\"app-logo-link\"]"
+    And I expect current url is "/"
```
Comment (justinehell) [inline cypress/e2e/HomePage/delete.project.feature:9]: jump line
`When I click on...`
`Then I expect current url...`

**Review by justinehell — CHANGES_REQUESTED** [review]

File: src/i18n/en-US/index.js [hunk 7]
```diff
@@ -18,6 +18,11 @@ export default {
       newProject: 'New project',
       importProject: 'Import project',
       importedProject: 'Project has been imported &#129395;!',
+      deleteProject: {
+        delete: 'Delete project',
+        confirmDelete: 'Are you sure you want to delete {name} ?',
```
Comment (justinehell) [inline src/i18n/en-US/index.js:23]: `{name}` -> `<b>{name}</b>`

**Review by justinehell — CHANGES_REQUESTED** [review]

File: cypress/e2e/HomePage/delete.project.feature [hunk 8]
```diff
@@ -0,0 +1,17 @@
+Feature: Test homepage: project deletion
+
+  Scenario: Delete existing project
```
Comment (Zorin95670) [inline cypress/e2e/HomePage/delete.project.feature:3]: Add new e2e test to delete project and recreate same project

**Comment by k-y0u** [comment]
- [ ] Change order of commits `Create DeleteProjectDialog` & `Create DeleteProjectForm`
- [ ] Put `Create ProjectEvent` before `Update deleteProjectById in Project`

File: src/pages/HomePage.vue (hunk 2 above)
Comment (justinehell) [reply src/pages/HomePage.vue:38]: don't think so, doc of the `getProjects()`  function says
`* @return {Object} Object that contains all project ids as keys and associated projects as values.`

File: src/composables/Project.js [hunk 9]
```diff
@@ -613,3 +602,24 @@ export async function initProject(project) {
...
+ */
+export async function deleteProjectById(projectId) {
+  const projects = getProjects();
+  const projectFiles = await readDir(`/${projectId}`);
+
+  if (projects[projectId]) {
+    projectFiles.forEach(async (file) => {
+      await deleteProjectFile(projectId, file, true);
+    });
+
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply src/composables/Project.js:617]: done

File: src/components/dialog/DeleteProjectDialog.vue [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+import DeleteProjectForm from 'src/components/form/DeleteProjectForm';
+import { onMounted, onUnmounted, ref } from 'vue';
+
+const projectId = ref(null);
+let dialogEventSubscription;
+
+/**
+ * Set projectId on valid event.
+ * @param {String} key - Event key.
+ * @param {Object} file - Event file.
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply src/components/dialog/DeleteProjectDialog.vue]: done

File: src/i18n/en-US/index.js [hunk 11]
```diff
@@ -18,6 +18,11 @@ export default {
...
+        deletedProject: 'Project has been deleted 😥',
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply src/i18n/en-US/index.js]: done

File: src/components/form/DeleteProjectForm.vue [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+import { ref } from 'vue';
+import { deleteProjectById } from 'src/composables/Project';
+import { useI18n } from 'vue-i18n';
+
+const emit = defineEmits(['project:delete']);
+
+const props = defineProps({
+  projectId: {
+    type: String,
+  },
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply src/components/form/DeleteProjectForm.vue]: done

File: src/components/form/DeleteProjectForm.vue [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+});
+
+const { t } = useI18n();
+const confirmDelete = ref(false);
+const submitting = ref(false);
+
+/**
+ * Delete project
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply src/components/form/DeleteProjectForm.vue]: done

File: cypress/e2e/HomePage/delete.project.feature (hunk 4 above)
Comment (Rahzbob) [reply cypress/e2e/HomePage/delete.project.feature]: done

File: cypress/e2e/HomePage/delete.project.feature (hunk 3 above)
Comment (Rahzbob) [reply cypress/e2e/HomePage/delete.project.feature]: done

File: cypress/e2e/HomePage/delete.project.feature (hunk 5 above)
Comment (Rahzbob) [reply cypress/e2e/HomePage/delete.project.feature]: done

File: cypress/e2e/HomePage/delete.project.feature (hunk 6 above)
Comment (Rahzbob) [reply cypress/e2e/HomePage/delete.project.feature]: done

File: cypress/e2e/HomePage/delete.project.feature (hunk 8 above)
Comment (Rahzbob) [reply cypress/e2e/HomePage/delete.project.feature]: done

File: src/components/card/ProjectCard.vue [hunk 14]
```diff
@@ -17,11 +17,28 @@
           {{ project.id }}
         </div>
       </q-img>
+
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply src/components/card/ProjectCard.vue]: done

File: src/components/form/DeleteProjectForm.vue [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+ */
+function onSubmit() {
+  submitting.value = true;
+  return deleteProjectById(props.projectId)
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply src/components/form/DeleteProjectForm.vue:59]: done

File: tests/unit/components/form/DeleteProjectForm.spec.js [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+    t: (t) => t,
+  }),
+}));
+
+jest.mock('src/composables/Project', () => ({
+  deleteProjectById: jest.fn((projectId) => {
+    if (projectId === 'error') {
+      return Promise.reject({ name: 'error' });
+    }
+    return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply tests/unit/components/form/DeleteProjectForm.spec.js:22]: done

File: tests/unit/components/form/DeleteProjectForm.spec.js [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+      props: {
+        projectId: 'test',
+      },
+    });
+  });
+
+  describe('Test variables initialization', () => {
+    describe('Test props: projectId', () => {
+      it('should match "projectId"', () => {
+        expect(wrapper.vm.props.projectId).toEqual('test');
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply tests/unit/components/form/DeleteProjectForm.spec.js]: done

File: tests/unit/components/form/DeleteProjectForm.spec.js [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+      await wrapper.vm.onSubmit();
+
+      expect(wrapper.vm.confirmDelete).toEqual(true);
+      expect(wrapper.emitted()['project:delete']).toBeTruthy();
+      expect(Notify.create).toHaveBeenCalledWith(expect.objectContaining({ type: 'positive' }));
+    });
+
+    it('should emit a negative notification on error', async () => {
+      await wrapper.setProps({ projectId: 'error' });
+      Notify.create = jest.fn();
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply tests/unit/components/form/DeleteProjectForm.spec.js:59]: done

File: tests/unit/components/dialog/DeleteProjectDialog.spec.js [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+    wrapper = shallowMount(DeleteProjectDialog, {
+      global: {
+        plugins: [
+          createI18n({ locale: 'en-US', messages: i18nConfiguration }),
+        ],
+      },
+    });
+  });
+
+  describe('Test function: setProjectId', () => {
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply tests/unit/components/dialog/DeleteProjectDialog.spec.js:37]: done

File: tests/unit/components/form/DeleteProjectForm.spec.js [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+
+  beforeEach(() => {
+    wrapper = shallowMount(DeleteProjectForm, {
+      props: {
+        projectId: 'test',
+      },
+    });
+  });
+
+  describe('Test variables initialization', () => {
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply tests/unit/components/form/DeleteProjectForm.spec.js:37]: done

File: tests/unit/components/form/DeleteProjectForm.spec.js [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+    describe('Test props: projectId', () => {
+      it('should match "projectId"', () => {
+        expect(wrapper.vm.props.projectId).toEqual('test');
+      });
+    });
+  });
+
+  describe('Test function: onSubmit', () => {
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply tests/unit/components/form/DeleteProjectForm.spec.js:45]: done

File: tests/unit/pages/HomePage.spec.js (hunk 1 above)
Comment (Rahzbob) [reply tests/unit/pages/HomePage.spec.js:38]: done

File: src/pages/HomePage.vue (hunk 2 above)
Comment (Rahzbob) [reply src/pages/HomePage.vue]: done

File: src/composables/Project.js [hunk 22]
```diff
@@ -613,3 +602,24 @@ export async function initProject(project) {
...
+ * Delete one project by ID.
+ * @param {String} projectId - Id of project.
+ * @return {Promise<void>} Promise with nothing on success otherwise an error.
+ */
+export async function deleteProjectById(projectId) {
+  const projects = getProjects();
+  const projectFiles = await readDir(`/${projectId}`);
+
+  if (projects[projectId]) {
+    projectFiles.forEach(async (file) => {
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply src/composables/Project.js]: done

File: src/composables/Project.js [hunk 23]
```diff
@@ -613,3 +602,24 @@ export async function initProject(project) {
...
+      await deleteProjectFile(projectId, file, true);
+    });
+
+    delete projects[projectId];
+    localStorage.setItem(PROJECT_STORAGE_KEY, JSON.stringify(projects));
+
+    ProjectEvent.DeleteProjectEvent.next();
```
[the remark this answers is not in the document]
Comment (Rahzbob) [reply src/composables/Project.js]: done

**Review by k-y0u — APPROVED** [review]

**Review by k-y0u — APPROVED** [review]

**Review by Zorin95670 — APPROVED** [review]

**Pull request merged**